src/lib/notifier.c

```diff
@@ -14,82 +14,163 @@
+	 *
+	 * Event producer might force unregister all receivers by passing
+	 * receiver NULL
+	 * Event consumer might unregister from all callers by passing caller
+	 * NULL
+	 */
 	spin_lock(notify->lock);
-	list_item_del(&notifier->list);
+	list_for_item_safe(wlist, tlist, &notify->list[type]) {
+		handle = container_of(wlist, struct callback_handle, list);
+		if ((!receiver || handle->receiver == receiver) &&
+		    (!caller || handle->caller == caller)) {
+			list_item_del(&handle->list);
+			rfree(handle);
+		}
+	}
 	spin_unlock(notify->lock);
 }
 
-void notifier_notify(void)
+void notifier_unregister_all(void *receiver, void *caller)
+{
+	int i;
+
+	for (i = 0; i < NOTIFIER_ID_MAX; i++)
+		notifier_unregister(receiver, caller, i);
+}
+
+static void notifier_notify(void *caller, enum notify_id type, void *data)
 {
 	struct notify *notify = *arch_notify_get();
 	struct list_item *wlist;
-	struct notifier *n;
-
-	if (!list_is_empty(&notify->list)) {
-		dcache_invalidate_region(&_notify_data, sizeof(_notify_data));
-		dcache_invalidate_region(_notify_data.data,
-					 _notify_data.data_size);
-
-		/* iterate through notifiers and send event to
-		 * interested clients
-		 */
-		list_for_item(wlist, &notify->list) {
-			n = container_of(wlist, struct notifier, list);
-			if (n->id == _notify_data.id)
-				n->cb(_notify_data.message, n->cb_data,
-				      _notify_data.data);
-		}
+	struct list_item *tlist;
+	struct callback_handle *handle;
+
+	/* iterate through notifiers and send event to
+	 * interested clients
+	 */
+	list_for_item_safe(wlist, tlist, &notify->list[type]) {
+		handle = container_of(wlist, struct callback_handle, list);
+		if (!caller || !handle->caller || handle->caller == caller)
+			handle->cb(handle->receiver, type, data);
 	}
 }
 
-void notifier_event(struct notify_data *notify_data)
+void notifier_notify_remote(void)
 {
 	struct notify *notify = *arch_notify_get();
+	struct notify_data *notify_data = &_notify_data[cpu_get_id()];
+
+	dcache_invalidate_region(notify_data, sizeof(*notify_data));
+	if (!list_is_empty(&notify->list[notify_data->type])) {
+		dcache_invalidate_region(notify_data->data,
+					 notify_data->data_size);
+		notifier_notify(notify_data->caller, notify_data->type,
+				notify_data->data);
+	}
+}
+
+void notifier_event(void *caller, enum notify_id type, uint32_t core_mask,
+		    void *data, uint32_t data_size)
+{
+	struct notify_data *notify_data;
 	struct idc_msg notify_msg = { IDC_MSG_NOTIFY, IDC_MSG_NOTIFY_EXT };
 	int i;
 
-	spin_lock(notify->lock);
-
-	_notify_data = *notify_data;
-	dcache_writeback_region(_notify_data.data, _notify_data.data_size);
-	dcache_writeback_region(&_notify_data, sizeof(_notify_data));
-
 	/* notify selected targets */
 	for (i = 0; i < PLATFORM_CORE_COUNT; i++) {
-		if (_notify_data.target_core_mask & (1 << i)) {
+		if (core_mask & NOTIFIER_TARGET_CORE_MASK(i)) {
 			if (i == cpu_get_id()) {
-				notifier_notify();
+				notifier_notify(caller, type, data);
```

Comment:
@mmaka1 I'm not sure if that's a good approach. To me it would look much better to have a wrapper with field specifying amount of elements in array, if you ever needed to pass an array via callback in the first place. Also, it would force all receivers to match signature with `data_size`, which would be unused in vast majority of cases.
If you insist I can add that, but I don't like it.